config/auth_metadata_service.go

```diff
@@ -0,0 +1,166 @@
+package config
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/databricks/databricks-sdk-go/logger"
+	"golang.org/x/oauth2"
+)
+
+// timeout to wait for metadata service to respond
+const metadataServiceTimeout = 10 * time.Second
+
+const MetadataServiceVersion = "1"
+const MetadataServiceVersionHeader = "X-Databricks-Metadata-Version"
+
+// response expected from the metadata service
+// other fields such as `cluster_id` might be added later
+type serverResponse struct {
+	// URL of the Databricks workspace to connect to
+	Host string `json:"host"`
+
+	// The requested access token.
+	Token oauth2.Token `json:"token"`
+}
+
+// Credentials provider that fetches a token from a locally running HTTP server
+//
+// The credentials provider will perform a GET request to the configured URL.
+//
+// The MUST return 4xx response if the "X-Databricks-Metadata-Version" header
+// is not set or set to a version that the server doesn't support.
+//
+// The server MUST guarantee stable sessions per URL path. That is, if the
+// server returns a token for a Host on a given URL path, it MUST continue to return
+// tokens for the same Host.
+//
+// The server is expected to return a JSON response with the following fields:
+//
+// host: URL of the Databricks host to connect to.
+// token: object with the following fields
+//   access_token: The requested access token.
+//   token_type: The type of token, which is a "Bearer" access token.
+//   expires_on: The timespan when the access token expires.
+
+type MetadataServiceCredentials struct{}
+
+func (c MetadataServiceCredentials) Name() string {
+	return "metadata-service"
+}
+
+func (c MetadataServiceCredentials) Configure(ctx context.Context, cfg *Config) (func(*http.Request) error, error) {
+	if cfg.MetadataServiceURL == "" {
+		return nil, nil
+	}
+
+	parsedMetadataServiceURL, err := url.Parse(cfg.MetadataServiceURL)
+	if err != nil {
+		return nil, fmt.Errorf("invalid auth server URL: %w", err)
+	}
+
+	// only allow localhost URLs
+	if parsedMetadataServiceURL.Hostname() != "localhost" && parsedMetadataServiceURL.Hostname() != "127.0.0.1" {
+		return nil, fmt.Errorf("invalid auth server URL: %s", cfg.MetadataServiceURL)
+	}
+
+	ms := metadataService{parsedMetadataServiceURL}
+
+	response, err := ms.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	if response == nil {
+		return nil, nil
+	}
+
+	// ignore this credential provider if it returns tokens for a different host
+	// if no host is configured yet then take the host from the metadata service
+	if cfg.Host != "" && cfg.Host != response.Host {
+		logger.Debugf(ctx, "ignoring metadata service because it returned a token for a different host: %s", response.Host)
+		return nil, nil
```

Comment:
@fjakobs This should return an error as well.

Reply:
For this I'm not so sure. Imagine you use host based auth and have a host configured. If a metadata service is available that can provide credentials for that host then it should be used but if the metadata service credentials are for another host then I think we should fall through and try the next method.

What do you think?